src/classes/GE_GiftEntryController.cls

```diff
@@ -60,63 +64,74 @@ public with sharing class GE_GiftEntryController {
     */
     @AuraEnabled
     public static String sendPurchaseRequest(String requestBodyParameters, String dataImportRecordId) {
+        UTIL_Http.Response response = processPurchaseRequest(requestBodyParameters, dataImportRecordId);
+
+        return JSON.serialize(response);
+    }
+
+    /***
+    * @description Sends the purchase request to Elevate and logs an error on a timeout error.
+    * This method is used in tests instead of the sendPurchaseRequest() in order to skip JSON.deserialize(response)
+    * due to it erroring out because the response header contains the Object data type.
+    * @param requestBodyParameters: JSON containing parameters for the purchase call request body
+    * @param dataImportRecordId: Record ID of the related Data Import
+    * @return UTIL_Http.Response
+    */
+    @TestVisible
+    private static UTIL_Http.Response processPurchaseRequest(String requestBodyParameters, String dataImportRecordId) {
+        UTIL_Http.Response response = new UTIL_Http.Response();
+
         try {
             GE_PaymentServices paymentServices = new GE_PaymentServices();
-            paymentServices.setPurchaseCallBody(requestBodyParameters);
-            GE_PaymentServices.ResponseWrapper response = paymentServices.createTransaction();
+            response = paymentServices.createTransaction(requestBodyParameters);
 
-            return JSON.serialize(response);
         } catch (Exception ex) {
-            Boolean isCalloutException = ex instanceof System.CalloutException;
-            Boolean hasTimedOut = isCalloutException && ex.getMessage().contains(TIMED_OUT);
+            String JSONExceptionData = ERR_ExceptionData.createExceptionWrapperJSONString(ex);
+            throw buildDmlException(JSONExceptionData);
 
-            if (hasTimedOut) {
-                return getPurchaseCallTimeoutResponse(ex, dataImportRecordId);
-            } else {
-                String JSONExceptionData = ERR_ExceptionData.createExceptionWrapperJSONString(ex);
-                throw returnDmlException(JSONExceptionData);
-            }
+        } finally {
+            processPurchaseTimeoutResponse(response, dataImportRecordId);
```

Comment:
Just to make sure I'm tracking this, these changes update the criteria we are using to identify a timed out response -  formerly the purchase was considered to have timed out when the CalloutException was thrown and the "timed out" text appeared in the message; after these changes it's looking at the response code of the response. I just want to make sure this shift is both intentional, and still in line with how we intended to process timeouts (CC @psadie-sfdo)

Reply:
The new [UTIL_Http.sendRequest()](https://github.com/SalesforceFoundation/NPSP/blob/70a0f7503bcc97841671f10542d4838c7e0e55da/src/classes/UTIL_Http.cls#L91) and the [timeout exception](https://github.com/SalesforceFoundation/NPSP/blob/70a0f7503bcc97841671f10542d4838c7e0e55da/src/classes/UTIL_Http.cls#L92) are created based on the original GE controller code. Difference is in the response `status` and `body` construction. The `sendRequest()` timeout error response `body` contains the exception error message, however, the GE controller builds the timeout response so that status contains the error message and the response `body` contains a generic timed out error. Thus, additional processing is done in the `processPurchaseTimeoutResponse()`.

So, `sendMessage()` sends the request and :
1.  if the Elevate returns a success response, the method just returns the response
2.  if the Elevate returns an error response, the method just returns the response
3.  if the Elevate returns a timed out response, the method just returns the response
4.  if the Callout exception is generated and the error message contains "timed out", a timed out response with status code 408, status "Request Timeout" and body the exception error message is returned.
5.  if any unexpected exception is generated and it is not a timed out, then a  bad request response is returned. I think this is different: a DML exception is returned currently in `master` on an unexpected send request exception.

GE `processPurchaseRequest()`:
- continues to build purchase request body and sends the purchase request.
- continues to throw DML exception on an unexpected error except in the above case (5)
- the format of the timed out error response and its conditions should not be changed.

Based on your comment, I thought of changing the code to not use `UTIL_Http.sendRequest()` but to use `HttpResponse requestResponse = new Http().send(request);` as before and built timed out response. But I think the only issue is the case (5), bad request response over DML exception when sending the request. Let me know what you think, I can update code not to use common `UTIL_Http.sendRequest()`. @psadie-sfdo , @patrick-yan-sf

Reply:
This looks good to me! I like UTIL_Http and would prefer that we keep using it moving forward. From what I'm seeing we're still logging the unexpected exceptions/errors in the Data Import record and also outputting an error to the UI.

The only thing I'd like to comment on here is with regards to `processPurchaseTimeoutResponse`. I'm not sure about always be making this call regardless of the status of the response. I would perhaps consider pulling the check on the response status code out and perform that check before calling `processPurchaseTimeoutResponse`.

`
if (response.statusCode == UTIL_Http.STATUS_CODE_REQUEST_TIMEOUT) {
                processPurchaseTimeoutResponse(response, dataImportRecordId);
            }
`

Reply:
Thanks for the feedback @psadie-sfdo ! I understand your point, previously the timeout error was processed on the callout exception but now it is every time on the purchase response validation. 

In the previous code version, the timeout as the result of the callout exception with the timed out message and the timeout returned from Elevate API differed, but now every timeout response (status code 408) will have status containing the error message and the response body containing the generic timed out error returned to the LWC. Does that present any issue?